Alter the code to work in terms of codepoints, not characters.

anarchodin opened the pull request:
I had a hunch about what would be causing the failures in ABCL and Allegro: They both operate using UTF-16 as an internal representation. I built [a small thing](https://github.com/anarchodin/trivial-utf-16) that turns Lisp strings into vectors of Unicode code points, whether the underlying implementation uses UTF-16 like ABCL or UTF-32 like SBCL. Then I changed uax-15 to work on UTF-32 rather than in terms of characters and added the new library as a dependency.

This eliminated the test failures on ABCL.

I'm not an experienced programmer, and suspect that there are tons of optimisation to be done, but this suggests I was right about the problem.

anarchodin committed:
Alter the code to work in terms of codepoints, not characters.

This adds a dependency on an external library, which handles the transformation
from Lisp strings to code point vectors. It likely causes non-trivial slowdown,
but that hasn't been measured at this point.

These changes result in a 100% pass rate for the tests in ABCL. It likely does
the same for Allegro, but has not been tested.

sabracrolleton commented:
Thank you very much for the proposed contribution. 

The good news, yes abcl and  allegro complete the tests with no failures.

The bad news: sbcl, ccl and ecl are slowed down; in the case of ccl it took more than three times longer to complete 10 runs of the test suite. 

Changes I would like to see to your pull request:
1. please include the trivial-utf-16 code as part of the uax-15 package.
2. make your changes as conditional compiles - e.g #-(or abcl allegro), #+(or abcl allegro)
so that sbcl, ccl and ecl are not slowed down while we think about possible optimization.

Thank you again for identifying a solution to the abcl and allegro problem.

Sabra

anarchodin commented:
> The bad news: sbcl, ccl and ecl are slowed down; in the case of ccl it took more than three times longer to complete 10 runs of the test suite.

A slowdown was likely. Experimenting just now, adding an inline declamation for the `to-unicode-string` and `from-unicode-string` seems to bring SBCL back to the same speed. (Although using conditional compilation to just skip straight to SBCL's own normalisation routines is still a _lot_ faster.) Doesn't do a blessed thing in CCL, though. I wonder if there are some particular requirements for inlining there that I'm not aware of.

>     2. make your changes as conditional compiles - e.g #-(or abcl allegro), #+(or abcl allegro)
>        so that sbcl, ccl and ecl are not slowed down while we think about possible optimization.

The problem is that the change is rather pervasive. I'll see if I can keep the conditionals localised rather than dotting them throughout.

anarchodin committed:
Embedded trivial-utf-16.

This is a little unclean, as the test also needs to do the transform, but the
functionality is well out of scope for uax-15 and is thus not exported.

anarchodin commented:
While looking for ways to keep the conditionals contained, I noticed that CCL's `map` was taking a full two orders of magnitude longer than in SBCL. I replaced the map, and my testing seems to indicate the slowdown is gone, with no change to anything but the trivial-utf-16 code. At that point I stopped thinking about how to keep the processing working on two different data types.

sabracrolleton commented:
Sounds good. I will validate on my side and then accept. I may try to do some testing on cmucl as well. Cmucl was choking on creating the unicode data  on compilation and never even got to the tests even if I applied your trivial-utf-16 code. Unfortunately I have not had time to try to figure that out.

anarchodin commented:
Checked it. I think I just completely skipped the QC bits; there was a `code-char` hiding out there which CMUCL choked on. Removing that makes the tests complete. CMUCL is working with UTF-16, and apparently does a type assertion on the argument of `code-char`.

This is non-standard behaviour, as `code-char` is supposed to return `NIL` if given a code that can't be turned into a character. Notably, there's another bugged system here: ABCL silently drops the leading bits. Allegro seems to do the right thing ... which, of course, means it won't complain and instead just stuffs a bunch of `NIL`s into the illegal lists. Both of those should do the right thing by not trying to convert the code to a character.

anarchodin committed:
Remove one last call to code-char.

This call caused a compilation error in CMUCL, and also causes the code not to
do the right thing for other 16-bit implementations.

sabracrolleton commented:
If you would like to be added to the list of authors, send me an email with preferred name and whether you want an email address included in the asd file.

anarchodin commented:
I meant to reply to this, but apparently haven't. I don't think there's any need for that.

Might be worth revisiting the README regarding those test failures, though. :)
